test/infrastructure/vcsim/controllers/virtualmachine_controller.go

```diff
@@ -0,0 +1,326 @@
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, errors.Wrap(err, "failed to get conditionsTracker")
+		}
+
+		conditionsTracker = &infrav1.VSphereVM{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      virtualMachine.Name,
+				Namespace: virtualMachine.Namespace,
+			},
+		}
+		if err := inmemoryClient.Create(ctx, conditionsTracker); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to create conditionsTracker")
+		}
+	}
+
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(virtualMachine, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Always attempt to Patch the VSphereVM + conditionsTracker object and status after each reconciliation.
+	defer func() {
+		// NOTE: Patch on VSphereVM will only add/remove a finalizer.
+		if err := patchHelper.Patch(ctx, virtualMachine); err != nil {
+			log.Error(err, "failed to patch VSphereVM")
+			if reterr == nil {
+				reterr = err
+			}
+		}
+
+		// NOTE: Patch on conditionsTracker will only track of provisioning process of the fake node, etcd, api server, etc.
+		if err := inmemoryClient.Update(ctx, conditionsTracker); err != nil {
+			log.Error(err, "failed to patch conditionsTracker")
+			if reterr == nil {
+				reterr = err
+			}
+		}
+	}()
+
+	// Handle deleted machines
+	if !vSphereMachine.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, r.reconcileDelete(ctx, cluster, machine, virtualMachine, conditionsTracker)
+	}
+
+	// Add finalizer first if not set to avoid the race condition between init and delete.
+	// Note: Finalizers in general can only be added when the deletionTimestamp is not set.
+	if !controllerutil.ContainsFinalizer(virtualMachine, VMFinalizer) {
+		controllerutil.AddFinalizer(virtualMachine, VMFinalizer)
+		return ctrl.Result{}, nil
+	}
+
+	// Handle non-deleted machines
+	return r.reconcileNormal(ctx, cluster, machine, virtualMachine, conditionsTracker)
+}
+
+func (r *VirtualMachineReconciler) reconcileNormal(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine, virtualMachine *vmoprv1.VirtualMachine, conditionsTracker *infrav1.VSphereVM) (ctrl.Result, error) {
+	ipReconciler := r.getVMIpReconciler(cluster, virtualMachine)
+	if ret, err := ipReconciler.ReconcileIP(ctx); !ret.IsZero() || err != nil {
+		return ret, err
+	}
+
+	bootstrapReconciler := r.getVMBootstrapReconciler(virtualMachine)
+	if ret, err := bootstrapReconciler.reconcileBoostrap(ctx, cluster, machine, conditionsTracker); !ret.IsZero() || err != nil {
+		return ret, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *VirtualMachineReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine, virtualMachine *vmoprv1.VirtualMachine, conditionsTracker *infrav1.VSphereVM) error {
+	bootstrapReconciler := r.getVMBootstrapReconciler(virtualMachine)
+	if ret, err := bootstrapReconciler.reconcileShutdown(ctx, cluster, machine, conditionsTracker); !ret.IsZero() || err != nil {
+		return err
```

Comment:
```suggestion
		return ret. err
```
Otherwise we just ignore !ret.IsZero